qiskit/circuit/library/boolean_logic.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=no-member
+
+"""Implementations of boolean logic quantum circuits.
+"""
+
+from typing import List, Optional
+
+import numpy as np
+from qiskit.circuit import QuantumRegister, QuantumCircuit
+from qiskit.circuit.exceptions import CircuitError
+
+
+class Permutation(QuantumCircuit):
+    """An n_qubit circuit that permutes qubits."""
+
+    def __init__(self,
+                 n_qubits: int,
+                 pattern: Optional[List[int]] = None,
+                 seed: Optional[int] = None) -> QuantumCircuit:
+        """Return an n_qubit permutation circuit impelemented using SWAPs.
+
+        Args:
+            n_qubits: circuit width.
+            pattern: permutation pattern. If None, permute randomly.
+            seed: random seed in case a random permutation is requested.
+
+        Returns:
+            A permutation circuit.
+
+        Raises:
+            CircuitError: if permutation pattern is malformed.
+        """
+        super().__init__(n_qubits, name="permutation")
+
+        if pattern is not None:
+            if sorted(pattern) != list(range(n_qubits)):
+                raise CircuitError("Permutation pattern must be some "
+                                   "ordering of 0..n_qubits-1 in a list.")
+            pattern = np.array(pattern)
+        else:
+            rng = np.random.RandomState(seed)
+            pattern = np.arange(n_qubits)
+            rng.shuffle(pattern)
+
+        for i in range(n_qubits):
+            if (pattern[i] != -1) and (pattern[i] != i):
```

Comment:
Users might try to specify the last index by using the `-1` index. It'd be safer to either disallow negative indices (less user friendly) or use some other means to check if the qubit has already been swapped, e.g. via a second list.

Reply:
yes good catch, the -1 shouldn't be there. i think it's safer to just require an actual 0..n-1 permutation, and I do raise if this is not the case (a few lines up).

Reply:
This -1 is just a flag to keep track of visited qubits. Any `pattern` which is not an exact permutation of [0, .., n-1] is explicitly disallowed a few lines up, so users putting -1 in there would see an error.

Reply:
Ah yes, and now it's also tested via the test you added 👍